pkg/controller/shoot/shoot_hibernation_control.go

```diff
@@ -0,0 +1,179 @@
+	return getShootHibernationSchedules(shoot) != nil
+}
+
+func (c *Controller) shootHibernationAdd(obj interface{}) {
+	if shootHasHibernationSchedules(obj.(*gardenv1beta1.Shoot)) {
+		key, err := cache.MetaNamespaceKeyFunc(obj)
+		if err != nil {
+			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+			return
+		}
+		c.shootHibernationQueue.Add(key)
+	}
+}
+
+func (c *Controller) shootHibernationUpdate(oldObj, newObj interface{}) {
+	var (
+		oldShoot = oldObj.(*gardenv1beta1.Shoot)
+		newShoot = newObj.(*gardenv1beta1.Shoot)
+
+		oldSchedule = getShootHibernationSchedules(oldShoot)
+		newSchedule = getShootHibernationSchedules(newShoot)
+	)
+
+	if !reflect.DeepEqual(oldSchedule, newSchedule) {
+		key, err := cache.MetaNamespaceKeyFunc(newObj)
+		if err != nil {
+			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
+			return
+		}
+		c.shootHibernationQueue.Add(key)
+	}
+}
+
+func (c *Controller) shootHibernationDelete(obj interface{}) {
+	if shootHasHibernationSchedules(obj.(*gardenv1beta1.Shoot)) {
+		key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+		if err != nil {
+			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+			return
+		}
+		c.shootHibernationQueue.Add(key)
+	}
+}
+
+func (c *Controller) deleteShootCron(logger logrus.FieldLogger, key string) {
+	if cr, ok := c.shootToHibernationCron[key]; ok {
+		cr.Stop()
+		logger.Debugf("Stopped cron")
+	}
+	delete(c.shootToHibernationCron, key)
+	logger.Debugf("Deleted cron")
+}
+
+func (c *Controller) reconcileShootHibernationKey(key string) error {
+	logger := hibernationLogger(key)
+	logger.Info("Shoot Hibernation")
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	shoot, err := c.shootLister.Shoots(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		c.deleteShootCron(logger, key)
+		logger.Debugf("Skipping because Shoot has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		logger.Debugf("Unable to retrieve object from store: %v", key, err)
+		return err
+	}
+
+	if shoot.DeletionTimestamp != nil {
+		c.deleteShootCron(logger, key)
+		return nil
+	}
+	return c.reconcileShootHibernation(logger, key, shoot.DeepCopy())
+}
+
+func (c *Controller) shootHibernationJob(logger logrus.FieldLogger, client garden.Interface, target *gardenv1beta1.Shoot, enabled bool) cron.Job {
+	return cron.FuncJob(func() {
+		_, err := kubernetes.TryUpdateShootHibernation(client, retry.DefaultBackoff, target.ObjectMeta,
+			func(shoot *gardenv1beta1.Shoot) (*gardenv1beta1.Shoot, error) {
+				if shoot.Spec.Hibernation == nil || !equality.Semantic.DeepEqual(target.Spec.Hibernation.Schedules, shoot.Spec.Hibernation.Schedules) {
+					return nil, fmt.Errorf("shoot %s/%s hibernation schedule changed mid-air", shoot.Namespace, shoot.Name)
+				}
+				shoot.Spec.Hibernation.Enabled = enabled
+				return shoot, nil
+			})
+		if err != nil {
```

Comment:
I suggest another check for `errors.IsNotFound` and ignore it afterwards. The map of schedules is accessed concurrently, thus there can be schedules for already deleted shoots.

Reply:
I'd rather keep the check for `IsNotFound` out of here since it keeps the logic and the concerns of a job simple.
Also, the map of schedules is accessed concurrently, yes, but for a single key it is only accessed once at a time due to the nature of the `workqueue`: It only allows a certain key to be processed once at a time. Thus, if you allowed the mutation of the schedules in the jobs you'd really have to start caring about concurrency.